From patchwork Mon Jul  8 16:47:51 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Wei Yang <w90p710@gmail.com>
X-Patchwork-Id: 11035465
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id C17FC1395
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  8 Jul 2019 16:48:00 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id AB321284C3
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  8 Jul 2019 16:48:00 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 9A287285C3; Mon,  8 Jul 2019 16:48:00 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No,
 score=-7.9 required=2.0 tests=BAYES_00,DKIM_ADSP_CUSTOM_MED,
	FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,UNPARSEABLE_RELAY
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 393B7284C3
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  8 Jul 2019 16:48:00 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1732123AbfGHQr7 (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 8 Jul 2019 12:47:59 -0400
Received: from out30-54.freemail.mail.aliyun.com ([115.124.30.54]:36564 "EHLO
        out30-54.freemail.mail.aliyun.com" rhost-flags-OK-OK-OK-OK)
        by vger.kernel.org with ESMTP id S1726318AbfGHQr7 (ORCPT
        <rfc822;kvm@vger.kernel.org>); Mon, 8 Jul 2019 12:47:59 -0400
X-Alimail-AntiSpam: 
 AC=PASS;BC=-1|-1;BR=01201311R101e4;CH=green;DM=||false|;FP=0|-1|-1|-1|0|-1|-1|-1;HT=e01e04407;MF=wei.yang1@linux.alibaba.com;NM=1;PH=DS;RN=2;SR=0;TI=SMTPD_---0TWPHB0h_1562604471;
Received: from localhost(mailfrom:wei.yang1@linux.alibaba.com
 fp:SMTPD_---0TWPHB0h_1562604471)
          by smtp.aliyun-inc.com(127.0.0.1);
          Tue, 09 Jul 2019 00:47:51 +0800
From: Wei Yang <w90p710@gmail.com>
To: kvm@vger.kernel.org
Cc: pbonzini@redhat.com
Subject: [PATCH] KVM: x86: Fix guest time accounting with
 VIRT_CPU_ACCOUNTING_GEN
Date: Tue,  9 Jul 2019 00:47:51 +0800
Message-Id: <20190708164751.88385-1-w90p710@gmail.com>
X-Mailer: git-send-email 2.14.1.40.g8e62ba1
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

move guest_exit() after local_irq_eanbled() so that the timer interrupt
hits we account that tick as spent in the guest.

Cc: Paolo Bonzini <pbonzini@redhat.com>
Signed-off-by: Wei Yang <w90p710@gmail.com>
---
 arch/x86/kvm/x86.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 2e302e977dac..04a2913f9226 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -8044,7 +8044,6 @@ static int vcpu_enter_guest(struct kvm_vcpu *vcpu)
 
 	++vcpu->stat.exits;
 
-	guest_exit_irqoff();
 	if (lapic_in_kernel(vcpu)) {
 		s64 delta = vcpu->arch.apic->lapic_timer.advance_expire_delta;
 		if (delta != S64_MIN) {
@@ -8054,6 +8053,7 @@ static int vcpu_enter_guest(struct kvm_vcpu *vcpu)
 	}
 
 	local_irq_enable();
+	guest_exit();
 	preempt_enable();
 
 	vcpu->srcu_idx = srcu_read_lock(&vcpu->kvm->srcu);
